### PR TITLE
Make it build with ghc-9.8

### PR DESCRIPTION
**Pull request by erikd**
# Description

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/shelley-ma/test-suite/cardano-ledger-shelley-ma-test.cabal
Comment (lehins) [inline eras/shelley-ma/test-suite/cardano-ledger-shelley-ma-test.cabal]: I suggest switching all packages to `<5` on `base`.

Comment (erikd) [reply eras/shelley-ma/test-suite/cardano-ledger-shelley-ma-test.cabal]: Done!

File: cabal.project
Comment (erikd) [inline cabal.project]: This is needed because partial functions like `head` and `tail` now emit a warning.

`ghc-options: -Werror` for all compiler versions will need to be restored once the code is fixed.

File: cabal.project
Comment (michaelpj) [inline cabal.project]: s-r-ps in conditionals don't work well, unfortunately.

I think you could just make this unconditional: it would be weird to have the ledger only sometimes depend on master of `plutus` for 9.8, in general that might not compile. Whereas making it always depend on master (for this branch which I assume you're not going to merge soon) is at least consistent.

File: .github/workflows/haskell.yml
Comment (lehins) [inline .github/workflows/haskell.yml]:
```suggestion
        ghc: ["8.10.7", "9.2.8", "9.6.3", "9.8.1"]
```

File: cabal.project
Comment (lehins) [reply cabal.project]: Why are we using SRP at all? The normal process is to make a releass of Plutus packages that are compatible with ghc-9.8 and then use plutus from CHaPs?

> I think you could just make this unconditional

It would create problems when we would release ledger packages to CHaPs. We can't have SRP on master at all.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Instead of using CPP to hide unzip, let's better make an explicit import of functions that are being used in this module from `Data.Functor`
```suggestion
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: CPP should be the last resort. In this case we can avoid it
```suggestion
```

File: libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs
Comment (lehins) [inline libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs]: Ditto
```suggestion
```

File: libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs
Comment (lehins) [inline libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs]:
```suggestion
```

File: cabal.project
Comment (lehins) [reply cabal.project]: Didn't you take care of the `head` and `tail` functions in this PR, or there are still some left over uses of those?

**Review by lehins — CHANGES_REQUESTED** [review]
As always, thank you for doing this compatibility work! We greatly appreciate it!

File: cabal.project
Comment (erikd) [reply cabal.project]: I made it unconditional.

If you want to do it without SRPs, you will have to wait until `ghc-9.8` compatible versions if the plutus packages hit CHaP.

**Comment by erikd** [comment]
> Didn't you take care of the head and tail functions in this PR, or there are still some left over uses of those?

@lehins I did not touch these functions because I thought I would leave it to the `ledger` maintainers to sort it out how they see fit.

File: .github/workflows/haskell.yml
Comment (erikd) [reply .github/workflows/haskell.yml]: Fixed.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs
Comment (erikd) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: CPP removed.

File: libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs
Comment (erikd) [reply libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs]: CPP removed.

File: libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs
Comment (erikd) [reply libs/ledger-state/src/Cardano/Ledger/State/UTxO.hs]: CPP removed.

**Comment by michaelpj** [comment]
Yeah, I doubt this needs to be merged soon, and in a week or so there will be a plutus release with some initial 9.8 support (still needing some allow-newers).

**Comment by erikd** [comment]
This PR no longer uses any SRPs other than `cardano-ledger-executable-spec`, so IMO is ready to be merged.

File: cabal.project
Comment (lehins) [inline cabal.project]: This misleading. It is not requirement to update both of the indices to get newest packages from hackage.
```suggestion
  -- Bump this if you need newer packages from Hackage
```

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Upec.hs
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Upec.hs]: Original implementation is pretty bad.  But the suggested one is not better either. Could you please switch back to `head` so we don't forget to fix it properly.
```suggestion
          (Just . applyPPUpdates pps . fst . head . Map.toList)
```

Proper fix is this:
```haskell
   in case Map.keys consensus of
        -- NOTE that `quorumN` is a global constant, and that we require
        -- it to be strictly greater than half the number of genesis nodes.
        -- The keys in the `pup` correspond to the genesis nodes,
        -- and therefore either:
        --   1) `consensus` is empty, or
        --   2) `consensus` has exactly one element.
        thePPups:_ -> Just $ applyPPUpdates pps thePPups
        -- NOTE that `updatePParams` corresponds to the union override right
        -- operation in the formal spec.
        _ -> Nothing
```

File: libs/cardano-ledger-test/cardano-ledger-test.cabal
Comment (lehins) [inline libs/cardano-ledger-test/cardano-ledger-test.cabal]: Omission of bounds here is intentional. We do not release `cardano-ledger-test` to CHaPs. So there is need to put whole bunch of unnecessary bounds in this package
```suggestion
        plutus-ledger-api,
```

File: eras/byron/chain/executable-spec/byron-spec-chain.cabal
Comment (lehins) [inline eras/byron/chain/executable-spec/byron-spec-chain.cabal:30]: Change like this requires a patch version update. Could you please bump up the version of `byron-spec-chain` in the cabal file and in the CHANGELOG from `1.0.0.1` to `1.0.0.2`

File: eras/byron/crypto/cardano-crypto-wrapper.cabal
Comment (lehins) [inline eras/byron/crypto/cardano-crypto-wrapper.cabal:76]: Change like this requires a patch version update. Could you please bump up the version of `byron-crypto-wrapper` in the cabal file from `1.5.1.0` to `1.5.1.1`

File: eras/byron/crypto/test/cardano-crypto-test.cabal
Comment (lehins) [inline eras/byron/crypto/test/cardano-crypto-test.cabal:54]: Change like this requires a patch version update. Could you please bump up the version of `cardano-crypto-test` in the cabal file from `1.5.0.0` to `1.5.0.1`

File: eras/byron/ledger/executable-spec/byron-spec-ledger.cabal
Comment (lehins) [inline eras/byron/ledger/executable-spec/byron-spec-ledger.cabal:39]: Change like this requires a patch version update. Could you please bump up the version of `byron-spec-ledger` in the cabal file and in the CHANGELOG from `1.0.0.1` to `1.0.0.2`

File: eras/conway/impl/cardano-ledger-conway.cabal
Comment (lehins) [inline eras/conway/impl/cardano-ledger-conway.cabal]: Where did `30` come from?
```suggestion
        base >=4.14 && <5,
```

File: libs/non-integral/non-integral.cabal
Comment (lehins) [inline libs/non-integral/non-integral.cabal:26]: Change like this requires a patch version update. Could you please bump up the version of `non-integral` in the cabal file from `1.0.0.0` to `1.0.0.1`

File: libs/set-algebra/set-algebra.cabal
Comment (lehins) [inline libs/set-algebra/set-algebra.cabal:33]: Change like this requires a patch version update. Could you please bump up the version of `set-algebra` in the cabal file from `1.1.0.1` to `1.1.0.2`

File: libs/small-steps-test/small-steps-test.cabal
Comment (lehins) [inline libs/small-steps-test/small-steps-test.cabal:33]: Change like this requires a patch version update. Could you please bump up the version of `small-steps-test` in the cabal file from `1.0.0.1` to `1.0.0.2`.

Also could I ask you to fix the changelog, it looks totally wrong right now:

```diff
-# Version history for `small-steps-test-test`
+# Version history for `small-steps-test`

-## 1.2.0.3
+## 1.0.0.2

-*
+* GHC-9.8 compatibility

-## 1.2.0.2
+## 1.0.0.1

-* GHC-9.6 ompatibility
+* GHC-9.6 compatibility
```

**Review by lehins — CHANGES_REQUESTED** [review]
Some of the changes are unnecessary and there are bunch of version bumps missing. I know that dealing with compatibility is not the most exciting part of writing programs, so thank you for doing this work!

File: libs/cardano-ledger-test/cardano-ledger-test.cabal
Comment (erikd) [reply libs/cardano-ledger-test/cardano-ledger-test.cabal]: Fixed.

File: libs/small-steps-test/small-steps-test.cabal
Comment (erikd) [reply libs/small-steps-test/small-steps-test.cabal:33]: Fixed.

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Upec.hs
Comment (erikd) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Upec.hs]: Reverted to the original code.

File: cabal.project
Comment (erikd) [reply cabal.project]: Fixed.

**Review by lehins — APPROVED** [review]
Beautiful. Thank you!

**Pull request merged**
